src/ukkonen.cpp

```diff
@@ -0,0 +1,936 @@
+      return to_human_readable_repr(self, ".");
+    });
+    state.def_readwrite("pos",
+                        &Ukkonen::State::pos,
+                        R"pbdoc(
+The position in the edge leading to the node ``v`` reached.)pbdoc");
+    state.def_readwrite("v",
+                        &Ukkonen::State::v,
+                        R"pbdoc(
+The index in :any:`Ukkonen.nodes` of the node at the end of the position reached.)pbdoc");
+    state.def(py::init<>(), R"pbdoc(
+Default constructor.)pbdoc");
+    state.def(py::init<node_index_type, edge_index_type>(),
+              py::arg("vv"),
+              py::arg("ppos"),
+              R"pbdoc(
+Construct from index and position.
+
+:param vv: the index of the node reached. 
+:type vv: int
+:param ppos: the position in the edge leading to vv.
+:type ppos: int
+)pbdoc");
+    state.def(py::init<Ukkonen::State const&>(), R"pbdoc(
+Default copy constructor.)pbdoc");
+    state.def("__copy__",
+              [](Ukkonen::State const& that) { return Ukkonen::State(that); });
+    state.def(py::self == py::self, py::arg("that"));
+
+    ////////////////////////////////////////////////////////////////////////
+    // Node
+    ////////////////////////////////////////////////////////////////////////
+
+    py::class_<Ukkonen::Node> node(uk,
+                                   "Node",
+                                   R"pbdoc(
+The type of the nodes in the tree.
+)pbdoc");
+    node.def("__repr__", [](Ukkonen::Node const& self) {
+      return to_human_readable_repr(self, ".");
+    });
+    node.def_readwrite("children",
+                       &Ukkonen::Node::children,
+                       R"pbdoc(
+The children of the current node.
+)pbdoc");
+    node.def_readwrite("l",
+                       &Ukkonen::Node::l,
+                       R"pbdoc(
+The index of the first letter in the edge leading to the node.
+)pbdoc");
+    node.def_readwrite("parent",
+                       &Ukkonen::Node::parent,
+                       R"pbdoc(
+The index of the parent node.
+)pbdoc");
+    node.def_readwrite("r",
+                       &Ukkonen::Node::r,
+                       R"pbdoc(
+The index of one past the last letter in the edge leading to the node.
+)pbdoc");
+    node.def_readwrite("link", &Ukkonen::Node::link);
+    node.def_readwrite("is_real_suffix", &Ukkonen::Node::is_real_suffix);
+    node.def(py::init<index_type, index_type, node_index_type>(),
+             py::arg("l")      = 0,
+             py::arg("r")      = 0,
+             py::arg("parent") = UNDEFINED,
+             R"pbdoc(
+Construct a node from left most index, right most index, and parent.
+
+:param l: the left most index and value of the data member l (defaults to ``0``). 
+:type l: int
+
+:param r: one after the right most index and value of the data member r (defaults to ``0``). 
+:type r: int
+
+:param parent: parent of the node being constructed (defaults to :any:`UNDEFINED`).
+:type parent: int
+)pbdoc");
```

Comment:
I'm having some issues with this definition when compiling with `clang`. The error I get is:
```cpp
src/ukkonen.cpp:628:32: error: use of overloaded operator '=' is ambiguous (with operand types 'py::arg' and 'const Undefined' (aka 'const Constant<0, detail::Max>'))
  628 |              py::arg("parent") = UNDEFINED,
      |              ~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~
```

I can make this go away by casting `UNDEFINED` to a `node_index_type` but this feels a bit dubious. What do you think @james-d-mitchell?

Reply:
I think casting to a `node_index_type` is safe here, it'll be passed as such to the underlying C++ constructor anyway, so might as well be statically cast here.